### DISCUSSION TITLE
Is it possible to make objects in an ActiveSelection not be positioned relative to selection?

### DISCUSSION THREAD

**Discussion by Naosyth** [fabricjs/fabric.js#9466] [Q&A] [answered]
For some context, I am migrating a codebase from v4 to v6, because we want to take advantage of some of the new changes regarding groups.

Our app has undo/redo logic that basically computes the difference in the state of the canvas after every modification, and uses that to step forward/backwards in our undo stack. The issue I'm running in to is that in our history stack, objects may or may not be recorded while part of an ActiveSelection, and also objects that are modified as a result of clicking undo or redo may _now_ be in an ActiveSelection, while they weren't when initially recorded. This results in positions often being set with the wrong frame of reference.

Is there any easy way to fix this, such as making ActiveSelection not reposition its children relative to itself?

**Comment by ShaMan123** [comment, marked as answer]
I have experienced the same.
Our fix was to check if the object is under an ActiveSelection and use `sendObjectToPlane` (applying or inverting the ActiveSelection transform).
That seems to be the best practice.

**Reply by lance-lfx** [reply]
I am running into a similar issue where I am trying to save the correct coordinates/location of the object based on the canvas plane when I have a modified event, and if I have an ActiveSelection, I am going through each of the object under ActiveSelection and saving their coordinates. However, obviously when I do that, the coordinates saved (ie. left or top) corresponds to the ActiveSelection plane and if I reload the object from the database, they would appear close to the origin (0,0) due to the fact that the object was saved with the coordinates of the ActiveSelection plane. So, I attempted to try what you suggested above:

`if (objects instanceof ActiveSelection) {
    objects.forEachObject(obj => {
        fabric.util.sendObjectToPlane(obj, objects.calcTransformMatrix());
        updateObjectInDatabase();
    }
}`

<img width="1014" alt="Screenshot 2024-10-17 at 11 17 32 AM" src="https://github.com/user-attachments/assets/ae801858-68a0-44c9-b424-e4a0e75fea1e">

The result of this is that it would update and save the object with the correct coordinates, however, on the canvas, the objects gets moved based on the new coordinates (coordinates of objects relative to the canvas). Until I reload the objects from the database, which would show them in the correct location relative to the canvas.

The image above shows the selection box of where it is supposed to be, while the objects are moved once I invoked the sendObjectToPlane function. The sendObjectToPlane description says that the canvas/viewer's perspective remains unchanged, but it seems like it does change.

Any suggestions to get around this?

**Reply by ShaMan123** [reply]
> if (objects instanceof ActiveSelection) { objects.forEachObject(obj => { fabric.util.sendObjectToPlane(obj, objects.calcTransformMatrix()); updateObjectInDatabase(); } }

Don't do that unless you undo it or remove the objects from the active selection or else you apply the transform matrix of the active selection twice on the objects.
Since you are persisting data use `sendPointToPlane` or `getXY`

**Reply by lance-lfx** [reply]
thanks for giving me the hint that it applies the transform matrix twice on the object, that makes a lot of sense
i got around it slightly differently, by getXY of the object under ActiveSelection and cloning the object and setting the cloned object's left/top with the Point from getXY and saving it to the database.

I tried removing the object from the ActiveSelection, but it didn't quite work, so I went for this alternate approach instead.

**Reply by ShaMan123** [reply]
Sounds bad actually.
I would avoid using a clone, and simply extract the data I want and send that over to the DB.

**Reply by saulacher** [reply]
I had the similar issue and solve it by adding obj.group = undefined; like this:
`if (objects instanceof ActiveSelection) { objects.forEachObject(obj => { fabric.util.sendObjectToPlane(obj, objects.calcTransformMatrix()); obj.group = undefined; updateObjectInDatabase(); } }`
